File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        int n = nums.length;
        int arr[] = new int[n+1];
        arr[0]=1;
        int count=0;
        int sum=0;
        for(int i:nums){
            sum+=i;
            if(sum>=goal){
                count+=arr[sum-goal];
            }
            arr[sum]++;
        }
        return count;
    }
}

File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.cpp
class Solution {
public:
    int helper(vector<int>& nums, int goal){
        if(goal<0) return 0;
        int count = 0;
        int sum = 0;
        int left = 0;
        int right = 0;
        while(right<nums.size()){
            sum+=nums[right];
            while(left<=right && sum>goal){
                sum = sum-nums[left];
                left++;
            }
            count+=right-left+1;
            right++;
        }
        return count;
    }
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        return helper(nums,goal)-helper(nums,goal-1);
    }
};